test/raft/distributed_test.cc

```diff
@@ -722,5 +722,87 @@ class delivery_queue {
     }
 };
 
+template <PureStateMachine M>
+struct raft_server {
+    raft::server_id _id;
+
+    std::unique_ptr<snapshots_t<typename M::state_t>> _snapshots;
+    std::unique_ptr<delivery_queue<typename M::state_t>> _queue;
+    std::unique_ptr<raft::server> _server;
+
+    // The following objects are owned by _server:
+    impure_state_machine<M>& _sm;
+    failure_detector& _fd;
+
+    // Points to a running fiber obtained by calling `_queue->receive_fiber()`.
+    // The fiber stops when we abort or crash.
+    std::optional<future<>> _msg_receiver;
+
+    // Advances the clocks of _server and _fd.
+    seastar::timer<lowres_clock> _ticker;
+};
+
+// Engaged optional: RPC message, nullopt: heartbeat
+template <typename State>
+using message_t = std::optional<typename rpc<State>::message_t>;
+
+template <PureStateMachine M>
+raft_server<M> create_server(raft::server_id id, network<message_t<typename M::state_t>>& net, bool first_server) {
+    using state_t = typename M::state_t;
+
+    auto snapshots = std::make_unique<snapshots_t<state_t>>();
+    auto sm = std::make_unique<impure_state_machine<M>>(*snapshots);
+    auto fd = seastar::make_shared<failure_detector>(
+            [id, &net] (raft::server_id dst) {
+        net.send(id, dst, std::nullopt);
+    });
+    auto rpc_ = std::make_unique<rpc<state_t>>(*fd, *snapshots,
+            [id, &net] (raft::server_id dst, typename rpc<state_t>::message_t m) {
+        net.send(id, dst, {std::move(m)});
+    });
+    auto persistence_ = std::make_unique<persistence<state_t, M::init>>(*snapshots, first_server ? std::optional{id} : std::nullopt);
+    auto queue = std::make_unique<delivery_queue<state_t>>(*rpc_);
+
+    auto& sm_ref = *sm;
+    auto& fd_ref = *fd;
+
+    auto server = raft::create_server(
+            id, std::move(rpc_), std::move(sm), std::move(persistence_), std::move(fd),
+            raft::server::configuration{});
+
+    return raft_server<M>{
+        ._id = id,
+        ._snapshots = std::move(snapshots),
+        ._queue = std::move(queue),
+        ._server = std::move(server),
+        ._sm = sm_ref,
+        ._fd = fd_ref
```

Comment:
fd is shared ptr, you can store ptr directly instead of a ref.

Reply:
Not sure if I want to do it: `raft_server::_fd` is supposed to only observe the failure detector, not own it; it's already owned by `raft::server` and we destroy `raft_server` before `raft::server`. Changing `raft_server::_fd` to a `shared_ptr` would unnecessarily extend ownership (I like to keep the number of owners of any resource to a minimum, preferably unique).

Reply:
OK